lib/CleanPlugin.js

```diff
@@ -0,0 +1,147 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Sergey Melyukov @smelukov
+*/
+
+"use strict";
+
+const path = require("path");
+const validateOptions = require("schema-utils");
+const schema = require("../schemas/plugins/CleanPlugin.json");
+
+/** @typedef {import("../declarations/plugins/CleanPlugin").CleanPluginArgument} CleanPluginArgument */
+/** @typedef {import("./Compiler")} Compiler */
+
+/**
+ * @param {import("./util/fs").OutputFileSystem} outputFS output fs
+ * @param {string} from from directory
+ * @returns {{files: Array<string>, directories: Array<string>}} collected files and directories
+ */
+function readDir(outputFS, from) {
+	const collectedFiles = [];
+	const collectedDirectories = [];
+	const stack = [from];
+	let cursor;
+
+	while ((cursor = stack.pop())) {
+		const stat = outputFS.statSync(cursor);
+
+		if (stat.isDirectory()) {
+			const items = outputFS.readdirSync(cursor);
+
+			if (from !== cursor) {
+				const relative = path.relative(from, cursor);
+				collectedDirectories.push(relative);
+			}
+
+			for (let i = 0; i < items.length; i++) {
+				stack.push(path.join(cursor, items[i]));
+			}
+		} else {
+			const relative = path.relative(from, cursor);
+			collectedFiles.push(relative);
+		}
+	}
+
+	return {
+		files: collectedFiles,
+		directories: collectedDirectories
+	};
+}
+
```

Comment:
> On incremental build it should not read from filesystem, but assume old state

There is `emittedOnce`-flag for that

Reply:
> There is `emittedOnce`-flag for that

ah I see. The flag causes the plugin to do nothing on incremental compilation, but instead it should e.g. remove assets from the first compilation that are no longer part of the second compilation.

I would solve that by storing the filesystem state after the first compilation (which directories now contain which files) and use that instead of reading the filesystem in the second compilation. As said we assume that nothing expect webpack changes the output directory while webpack is running. This should be pretty efficient, no fs calls should be done expect for removing old assets, when they are no longer referenced.